cmd/bad-key-revoker/main.go

```diff
@@ -0,0 +1,360 @@
+	initialID := 0
+	for {
+		var batch []struct {
+			ID         int
+			CertSerial string
+		}
+		_, err := bkr.dbMap.Select(
+			&batch,
+			"SELECT id, certserial FROM keyHashToSerial WHERE keyHash = ? and id > ? ORDER BY id LIMIT ?",
+			row.KeyHash,
+			initialID,
+			bkr.certificatesBatchSize,
+		)
+		if err != nil {
+			if db.IsNoRows(err) {
+				return unrevokedCerts, nil
+			}
+			return nil, err
+		}
+		for _, serial := range batch {
+			var unrevokedCert unrevoked
+			err = bkr.dbMap.SelectOne(
+				&unrevokedCert,
+				`SELECT cs.serial, c.registrationID, c.der FROM certificateStatus AS cs
+				JOIN certificates AS c
+				ON cs.serial = c.serial
+				WHERE cs.serial = ? AND cs.isExpired = false AND cs.status != ?`,
+				serial.CertSerial,
+				string(core.StatusRevoked),
+			)
+			if err != nil {
+				if db.IsNoRows(err) {
+					continue
+				}
+				return nil, err
+			}
+			unrevokedCert.RevokedBy = row.RevokedBy
+			unrevokedCerts = append(unrevokedCerts, unrevokedCert)
+		}
+		if len(batch) != bkr.certificatesBatchSize {
+			return unrevokedCerts, nil
+		}
+		initialID = batch[len(batch)-1].ID
+	}
+}
+
+func (bkr *badKeyRevoker) markRowChecked(row unchecked) error {
+	_, err := bkr.dbMap.Exec("UPDATE blockedKeys SET extantCertificatesChecked = true WHERE keyHash = ?", row.KeyHash)
+	return err
+}
+
+func (bkr *badKeyRevoker) resolveContacts(ids []int64) (map[int64]string, error) {
+	idToEmail := map[int64]string{}
+	for _, id := range ids {
+		var emails struct {
+			Contact []string
+		}
+		err := bkr.dbMap.SelectOne(&emails, "SELECT contact FROM registrations WHERE id = ?", id)
+		if err != nil {
+			return nil, err
+		}
+		if len(emails.Contact) == 0 {
+			continue
+		}
+		idToEmail[id] = strings.TrimPrefix(emails.Contact[0], "mailto:")
+	}
+	return idToEmail, nil
+}
+
+var emailTemplate = template.Must(template.New("email-template").Parse(`Hello,
+
+The public key associated with certificates which you have issued has been marked as compromised. As such we are required to revoke any certificates which contain this public key.
+
+The following currently unexpired certificates that you've issued contain this public key and have been revoked:
+{{range . -}}
+{{.}}
+{{end}}`))
+
+var emailSubject = "Certificates you've issued have been revoked due to key compromise"
+
+var maxSerials = 100
+
+func (bkr *badKeyRevoker) sendMessages(mapping map[string][]string) error {
+	err := bkr.mailer.Connect()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = bkr.mailer.Close()
+	}()
+	for email, serials := range mapping {
+		if len(serials) > maxSerials {
+			more := len(serials) - maxSerials
+			serials = serials[:maxSerials]
+			serials = append(serials, fmt.Sprintf("and %d more certificates.", more))
+		}
+		message := bytes.NewBuffer(nil)
+		err := emailTemplate.Execute(message, serials)
+		if err != nil {
+			return err
+		}
+		err = bkr.mailer.SendMail([]string{email}, emailSubject, message.String())
+		if err != nil {
```

Comment:
If one of the email sends fails, we should still go on and try the other emails. Consider, for example, the case where someone's email address is malformed (which we try to catch on ingestion and elsewhere, but we do not always catch).